File: src/libs/heap.py
from heapq import heappop, heappush


class Heap:
    def __init__(self):
        self.h = []

    def __len__(self):
        return len(self.h)

    def push(self, node):
        # The id(node) is used for cost tie-breaking
        heappush(self.h, (node.distance, id(node), node.index))

    def pop(self):
        return heappop(self.h)[-1]

File: src/libs/heap.cpp
#include "heap.hpp"

int Heap::Size()
{
    return heap.size();
}

void Heap::Push(int distance, int node_index)
{
    heap.push(std::make_pair(distance, node_index));
}

int Heap::Pop()
{
    std::pair<int, int> minNode = heap.top();
    heap.pop();

    return minNode.second;
}
